### PR TITLE
Set scope 

### LINKED ISSUE #46: Delete snippet if owner

### LINKED ISSUE #155: fix manager dashboard layout on page view
negative margin-left...

### LINKED ISSUE #166: Add toasts for various actions

### LINKED ISSUE #182: Fix $watch function on dashboard so only ever one assigned and always with the current $scope instance

### LINKED ISSUE #190: Smooth out shoddy state transitions

### PR DESCRIPTION
**Pull request by ctheller**
closes  #182 and closes #166 and closes #181 and closes #155  @omriBernstein also closes #190 closes #46 closes #93

### CODE REVIEW

**Comment by omriBernstein** [comment]
That's a whole lot of closes!

**Comment by omriBernstein** [comment]
Broader comment about commit messages: 1) nice job keeping them short but still descriptive 2) it's generally considered best practice to use present-tense imperative instead of past tense. So "fix state changes" as opposed to "fixed state changes". You can [read more here](http://chris.beams.io/posts/git-commit/#imperative).

File: browser/js/app.js [hunk 1]
```diff
@@ -32,6 +32,14 @@ app.config(function($urlRouterProvider, $locationProvider, $mdThemingProvider, $
 // This app.run is for controlling access to specific states.
 app.run(function($rootScope, AuthService, $state) {
 
+    // $rootScope.$on('$stateChangeStart', function(event, toState, toParams, fromState, fromParams, options){
```
Comment (omriBernstein) [inline browser/js/app.js]: Dead code should be given a proper burial.

File: browser/js/app.js [hunk 2]
```diff
@@ -65,4 +73,7 @@ app.run(function($rootScope, AuthService, $state) {
 
     });
 
+    //add lodash to angular view
+    $rootScope._ = window._;
```
Comment (omriBernstein) [inline browser/js/app.js]: Unless you're using `_` all over your views, I'd recommend making `_` a value provider instead of a property on the `$rootScope`. Something like `app.value('_', window._)` and then you can inject it into other factories, controllers, etc as needed.

File: browser/js/common/directives/snippet/snippet.html [hunk 3]
```diff
@@ -15,7 +15,7 @@
             <span ng-click="removeCollaborator(collaborator.$id)" class="close"><i class="material-icons"> clear</i></span>
 
             <md-tooltip>{{collaborator.first_name}} {{collaborator.last_name}}</md-tooltip>
-            <img ng-src="{{ collaborator.photoUrl || '/files/default-profile.png' }}" style="z-index: {{ collaborators.length - index }}">
+            <img ng-src="{{ collaborator.photoUrl || '/files/default-profile.png' }}" style="z-index: {{ collaborators.length - index }}" ng-cloak>
```
Comment (omriBernstein) [inline browser/js/common/directives/snippet/snippet.html]: You might consider creating users instances with a default `photoUrl`—like stored in the db.

Comment (omriBernstein) [inline browser/js/common/directives/snippet/snippet.html]: There's also `ng-style` by the way. So in this case you could instead have `ng-style="{'z-index': collaborators.length - index}"`.

File: browser/js/common/directives/snippet/snippet.html [hunk 4]
```diff
@@ -45,7 +45,7 @@
                 <div class="snippetImages img-wrap" ng-repeat="(index,collaborator) in collaborators track by $index">
                     <span ng-show="type!=='report'" ng-click="removeCollaborator(collaborator.$id)" class="close expansion"><i class="material-icons"> clear</i></span>
                     <md-tooltip>{{collaborator.first_name}} {{collaborator.last_name}}</md-tooltip>
-                    <img class="collaborator-avatar" ng-src="{{ collaborator.photoUrl || '/files/default-profile.png' }}">
+                    <img class="collaborator-avatar" ng-src="{{ collaborator.photoUrl || '/files/default-profile.png' }}" ng-cloak>
```
Comment (omriBernstein) [inline browser/js/common/directives/snippet/snippet.html]: Yeah if you find yourself repeating this a lot it could make a lot of sense to just always store a default value with the instance.

File: browser/js/common/directives/snippet/snippetDirective.js [hunk 5]
```diff
@@ -28,16 +28,15 @@ app.directive('snippet', function($rootScope, $state, Snippet, $mdExpansionPanel
...
 
             Snippet.getSnippetById(scope.id).$bindTo(scope, 'snippet');
 
             //ng-repeat through object directly instead!!
-            scope.$watch('snippet', function(newValue, oldValue) {
-                if (newValue)
-                    var i = 0;
+            scope.$watch('snippet', function() {
+                
+                var i = 0;
```
Comment (omriBernstein) [inline browser/js/common/directives/snippet/snippetDirective.js]: I don't see `i` being used below.

File: browser/js/common/directives/snippet/snippetDirective.js [hunk 6]
```diff
@@ -51,6 +50,7 @@ app.directive('snippet', function($rootScope, $state, Snippet, $mdExpansionPanel
             scope.plusButton = 'http://joshiscorner.com/files/images/plusButton.png';
 
             scope.collapse = function() {
+                console.log(scope.id, scope.type);
```
Comment (omriBernstein) [inline browser/js/common/directives/snippet/snippetDirective.js]: 🔥  🌳 🔥

File: browser/js/common/factories/snippet-factory.js [hunk 7]
```diff
@@ -84,17 +84,17 @@ app.factory("Snippet", function($firebaseObject, AuthService, Users, $rootScope,
...
             var dataToTransfer = { subject: fromSnippet.subject, contents: fromSnippet.contents };
-            Snippet.create(dataToTransfer);
+            return Snippet.create(dataToTransfer);
         })
 
     };
 
 
     Snippet.delete = function(snippetId) {
+        console.log(snippetId);
```
Comment (omriBernstein) [inline browser/js/common/factories/snippet-factory.js]: You could build a cabin with all these :)

File: browser/js/common/factories/snippet-factory.js [hunk 8]
```diff
@@ -106,12 +106,45 @@ app.factory("Snippet", function($firebaseObject, AuthService, Users, $rootScope,
             })
             updates[`/users/${snippet.team}/snippets/asManager/${snippetId}`] = null;
             updates[`/users/${snippet.owner}/snippets/asOwner/${snippetId}`] = null;
+            updates[`/users/${snippet.owner}/snippets/asCollaborator/${snippetId}`] = null;
```
Comment (omriBernstein) [inline browser/js/common/factories/snippet-factory.js]: Another neat es6 trick. You can embed `[]` for an object literal *key* to specify a dynamic value. So this could be ``updates = {[`/users/${snippet.owner}/snippets/asCollaborator/${snippetId}`]: null}``. Not at all necessary, but thought it would be useful info for you all.

File: browser/js/common/factories/snippet-factory.js [hunk 9]
```diff
@@ -106,12 +106,45 @@ app.factory("Snippet", function($firebaseObject, AuthService, Users, $rootScope,
...
             updates[`/users/${snippet.owner}/snippets/asTeamMember/${snippetId}`] = null;
             updates['/snippets/' + snippetId] = null;
-            return ref.update(updates);
+            console.log('setting the updates');
+            return ref.update(updates).then(function(result){
+                return result;
```
Comment (omriBernstein) [inline browser/js/common/factories/snippet-factory.js]: Unnecessary return. The promise is already a promise for the result, prior to the `.then`.

File: browser/js/common/factories/snippet-factory.js [hunk 10]
```diff
@@ -106,12 +106,45 @@ app.factory("Snippet", function($firebaseObject, AuthService, Users, $rootScope,
             })
             updates[`/users/${snippet.team}/snippets/asManager/${snippetId}`] = null;
             updates[`/users/${snippet.owner}/snippets/asOwner/${snippetId}`] = null;
+            updates[`/users/${snippet.owner}/snippets/asCollaborator/${snippetId}`] = null;
             updates[`/users/${snippet.owner}/snippets/asTeamMember/${snippetId}`] = null;
             updates['/snippets/' + snippetId] = null;
-            return ref.update(updates);
+            console.log('setting the updates');
```
Comment (omriBernstein) [inline browser/js/common/factories/snippet-factory.js]: ~~`console.log('setting the updates');`~~

File: browser/js/dashboard/dash.wrapper.controller.js [hunk 11]
```diff
@@ -19,4 +26,9 @@ app.controller('DashWrapperCtrl', function($scope, $state, $mdDialog, MdHelpers)
         });
     };
 
+    $scope.$on('$stateChangeStart', function(){
+        $element.remove();
```
Comment (omriBernstein) [inline browser/js/dashboard/dash.wrapper.controller.js]: Why is this needed?

File: browser/js/dashboard/dashboard.controller.js [hunk 12]
```diff
@@ -4,48 +4,33 @@ app.controller('DashboardCtrl', function($rootScope, $scope, $mdDialog, MdHelper
         $scope.draggingNow = true;
     }
 
-    var dateFilter = function(obj){
-        return _.pickBy(obj, function(snippetCreated){
-            var diff = (snippetCreated - $scope.currentWeek)/(1000*60*60*24);
-            return (diff < 7 && diff > 0);
-        })
+    $scope.dateInRange = function(date){
```
Comment (omriBernstein) [inline browser/js/dashboard/dashboard.controller.js]: This could be a good place for your own [custom angular filter](https://scotch.io/tutorials/building-custom-angularjs-filters).

File: browser/js/dashboard/dash.wrapper.controller.js (hunk 11 above)
Comment (ctheller) [inline browser/js/dashboard/dash.wrapper.controller.js]: Our state changes are (for some reason) not removing the old state from the
DOM until about a second after the transition. So I had to add these to
force clear the DOM before the new state loads (possibly because of one of
the animation modules we injected?? Idk. I troubleshooted this for a
whiiiile).

Chris Heller
781-718-7743
www.chrishellermusic.com

File: browser/js/dashboard/dashboard.html [hunk 13]
```diff
@@ -1,51 +1,49 @@
...
-    <div class="report-container" layout="column" flex="50" ng-if="isManager">
+    <div class="report-container" layout="column" flex="50" ng-if="user.reports">
         <div class="report-title" layout="row" ng-cloak>
             Your reports
         </div>
         <md-expansion-panel-group md-component-id="report-panel" multiple>
-            <div class="parent" ng-repeat="obj in reportSnippetIds" layout="row" data-drag="true" ng-model="reportSnippetIds" data-jqyoui-options="{revert: 'invalid', helper: 'clone'}" jqyoui-draggable="{index: {{$index}}, placeholder: 'keep'}">
-                <snippet class="fromReport" id="obj.id" type="obj.type" card="false"></snippet>
+            <div ng-if="dateInRange(value)" class="parent" ng-repeat="(key, value) in user.snippets.asManager" layout="row" data-drag="true" ng-model="user.snippets.asManager" data-jqyoui-options="{revert: 'invalid', helper: 'clone'}" jqyoui-draggable="{index: {{$index}}, placeholder: 'keep'}">
+                <snippet class="fromReport" id="key" type="'report'" card="false"></snippet>
```
Comment (omriBernstein) [inline browser/js/dashboard/dashboard.html]: Beware id inside an `ng-repeat`. Here multiple elements will end up with the same id.

File: browser/js/search-popup/search-popup.html [hunk 14]
```diff
@@ -9,7 +9,7 @@
   <md-dialog-content style="width: 600px">
     <div style="width: 600px" id="search-popup" class="md-dialog-content">
       <form style="width: 400px" ng-submit="$event.preventDefault()">
-        <md-autocomplete
+        <md-autocomplete
```
Comment (omriBernstein) [inline browser/js/search-popup/search-popup.html]: +1

File: browser/js/dashboard/dashwrapper.html [hunk 15]
```diff
@@ -1,11 +1,11 @@
...
+<div layout="row" class="dashboard-wrapper" ng-cloak>
     <div layout="column" flex style="height: 100%">
-        <div layout="row" id="dateNav">
-            <a ui-sref="dashboard.week({week:currentWeekNum-1})" ui-sref-opts="{reload: true, inherit: false}">
+        <div ng-if="loaded" layout="row" id="dateNav">
+            <a ng-click="go(currentWeekNum-1)">
                 <i class="left-one material-icons">keyboard_arrow_left</i>
             </a>
             <span id="displayweek" ng-cloak>{{displayWeek}}</span>
-            <a ui-sref="dashboard.week({week:currentWeekNum+1})" ui-sref-opts="{reload: true, inherit: false}">
```
Comment (omriBernstein) [inline browser/js/dashboard/dashwrapper.html]: Switching from the `ui-sref` means users won't be able to copy the url, nor see the url (in the bottom left of the browser window) when hovering over it. Still might make sense to keep this decision, but just thought I'd mention.

File: browser/js/dashboard/dashboard.controller.js [hunk 16]
```diff
@@ -4,48 +4,33 @@ app.controller('DashboardCtrl', function($rootScope, $scope, $mdDialog, MdHelper
...
+        return ((date - $scope.currentWeek)/(1000*60*60*24) < 7 && (date - $scope.currentWeek)/(1000*60*60*24) > 0);
     }
 
     var setScope = function(){
-        if (!$rootScope.user.snippets) {
-            $scope.teamSnippetIds = $scope.collabSnippetIds = $scope.collabAndTeamSnippetIds = $scope.reportSnippetIds = $scope.mySnippetIds = $scope.allSnippetIds = [];
-            return;
-        }
-        $scope.mySnippetIds = $rootScope.user.snippets.asOwner ? Object.keys(dateFilter($rootScope.user.snippets.asOwner)) : [];
```
Comment (omriBernstein) [inline browser/js/dashboard/dashboard.controller.js]: Nice. It *had* to feel good to refactor this.

File: browser/js/search-popup/search-popup.js [hunk 17]
```diff
@@ -91,7 +87,7 @@ function DialogCtrl ($timeout, $q, $scope, $mdDialog, $rootScope, Users) {
       var lowercaseQuery = angular.lowercase(query);
 
       return function filterFn(employee) {
-        return (employee.value.indexOf(lowercaseQuery) === 0);
+        return (employee.value.indexOf(lowercaseQuery) !== -1);
```
Comment (omriBernstein) [inline browser/js/search-popup/search-popup.js]: A new friend for you [es6 `includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes).

**Comment by omriBernstein** [comment]
@ctheller and team: I added my two cents. Feel free to add your own and/or ask questions here, too.
